third_party/exprtk/exprtk.hpp

```diff
@@ -3188,7 +3188,10 @@ namespace exprtk
                }
             }
 
-            return (g.token_list_.size() - stride_ + 1);
+            // This return value appears to be unused and may cause undefined
+            // behavior when stride_ > (g.token_list_.size() + 1)
+            // return (g.token_list_.size() - stride_ + 1);
+            return 0;
```

Comment:
@ArashPartow . Here's a minimal example: 

test.cpp

```
#include "../gdal/third_party/exprtk/exprtk.hpp"

int main() {
    exprtk::expression<double> expression;
    exprtk::parser<double> parser;
    exprtk::symbol_table<double> table;

    double x = 8;
    table.add_variable("X", x);
    expression.register_symbol_table(table);
    parser.compile("X", expression);

    double value = expression.value();

    return value != 8;
}
```
running

```
clang++ -fsanitize=undefined -fsanitize=unsigned-integer-overflow test.cpp && ./a.out
```

produces
```
../gdal/third_party/exprtk/exprtk.hpp:3193:42: runtime error: unsigned integer overflow: 1 - 2 cannot be represented in type 'unsigned long'
SUMMARY: UndefinedBehaviorSanitizer: undefined-behavior ../gdal/third_party/exprtk/exprtk.hpp:3193:42 in 
../gdal/third_party/exprtk/exprtk.hpp:3193:52: runtime error: unsigned integer overflow: 18446744073709551615 + 1 cannot be represented in type 'unsigned long'
SUMMARY: UndefinedBehaviorSanitizer: undefined-behavior ../gdal/third_party/exprtk/exprtk.hpp:3193:52 in
```

Originally showed up in our CI here (https://github.com/OSGeo/gdal/actions/runs/11784348075/job/32823345209#step:17:29078)

Reply:
@dbaston  I have been able to replicate that sanitizer diagnostic.

Initially, in C++, arithmetic operations between unsigned integer types cannot either overflow or underflow, hence there is no undefined behaviour. In short this particular check is broken, and that is probably why it is not on/enabled by default when enabling UBSAN.

The for the sequence of operations should be:

```
 1. uint - uint     -> uint
 2. uint + int
    2.1 uint + uint -> uint
```


In any case people more familiar than myself seem to think as much: https://gcc.gnu.org/bugzilla/show_bug.cgi?id=106469

I suspect `"sanitize=unsigned-integer-overflow"` is similar to many other checks that can be enabled, where these checks are better off being used periodically on a developer's box than to have them be run on CI continuously.


That being said, the return values in that expression stopped being used a long time ago, the operator calls now register the encountered errors internally in the `token_scanner`.  I've updated the library, where the change is to set all the return values to be zero, a further refactoring may be done down the road to remove the return values altogether. This change remove the diagnostic that is raised. Let me know if this fixes the issue you're seeing.

Reply:
> Initially, in C++, arithmetic operations between unsigned integer types cannot either overflow or underflow, hence there is no undefined behaviour.

They actually overflow or underflow (ie UINT_MAX + 1 == 0), but that is indeed defined behavior. That said, my experience with the GDAL library is that 75% of the situations where unsigned integer overflow/underflow happened were actual bugs, hence we enable it in our UBSAN continuous integration.  We have a special function decorator ``#define CPL_NOSANITIZE_UNSIGNED_INT_OVERFLOW     __attribute__((no_sanitize("unsigned-integer-overflow")))`` to tag the few functions where unsigned integer overflow/underflow is the desired behavior  (like when doing a horizontal difference filter for integer data)

Reply:
@rouault  Just to clarify my statement about UBSAN: is that it should be on (used) and the default set of checks are fine. 

The UBSAN checks that are not on by **default**, are probably so because they are problematic in certain contexts giving false positives and require more attention be provided when analysing their results. 

That being said, if defined behaviour (in this case of wrap around post unsigned operations) provides incorrect results, that is most likely a bug in the code. _As an example the wheels on a car rotating (aka behaving correctly) are not the cause of the crash the car might be in._